File: open-nucleus-app/lib/core/constants/permissions.dart
/// RBAC permissions and role definitions mirroring the Go backend
/// (`internal/model/rbac.go`).
///
/// The [rolePermissions] map is the single source of truth on the Flutter side
/// and **must** stay in sync with the backend whenever roles or permissions
/// change.
class Permissions {
  Permissions._();

  // ---------------------------------------------------------------------------
  // Permission constants
  // ---------------------------------------------------------------------------

  static const String patientRead = 'patient:read';
  static const String patientWrite = 'patient:write';
  static const String encounterRead = 'encounter:read';
  static const String encounterWrite = 'encounter:write';
  static const String observationRead = 'observation:read';
  static const String observationWrite = 'observation:write';
  static const String conditionRead = 'condition:read';
  static const String conditionWrite = 'condition:write';
  static const String medicationRead = 'medication:read';
  static const String medicationWrite = 'medication:write';
  static const String allergyRead = 'allergy:read';
  static const String allergyWrite = 'allergy:write';
  static const String conflictRead = 'conflict:read';
  static const String conflictResolve = 'conflict:resolve';
  static const String alertRead = 'alert:read';
  static const String alertWrite = 'alert:write';
  static const String syncRead = 'sync:read';
  static const String syncTrigger = 'sync:trigger';
  static const String formularyRead = 'formulary:read';
  static const String formularyWrite = 'formulary:write';
  static const String anchorRead = 'anchor:read';
  static const String anchorTrigger = 'anchor:trigger';
  static const String supplyRead = 'supply:read';
  static const String supplyWrite = 'supply:write';
  static const String smartLaunch = 'smart:launch';
  static const String smartRegister = 'smart:register';
  static const String deviceManage = 'device:manage';
  static const String consentRead = 'consent:read';
  static const String consentWrite = 'consent:write';

  // ---------------------------------------------------------------------------
  // Role constants
  // ---------------------------------------------------------------------------

  static const String roleCHW = 'community_health_worker';
  static const String roleNurse = 'nurse';
  static const String rolePhysician = 'physician';
  static const String roleSiteAdmin = 'site_administrator';
  static const String roleRegionalAdmin = 'regional_administrator';

  static const List<String> allRoles = [
    roleCHW,
    roleNurse,
    rolePhysician,
    roleSiteAdmin,
    roleRegionalAdmin,
  ];

  // ---------------------------------------------------------------------------
  // Role -> permissions mapping (mirrors internal/model/rbac.go exactly)
  // ---------------------------------------------------------------------------

  static const Map<String, List<String>> rolePermissions = {
    roleCHW: [
      patientRead,
      observationRead,
      observationWrite,
      alertRead,
      syncRead,
    ],
    roleNurse: [
      patientRead,
      encounterRead,
      encounterWrite,
      observationRead,
      observationWrite,
      medicationRead,
      conditionRead,
      allergyRead,
      alertRead,
      syncRead,
    ],
    rolePhysician: [
      patientRead,
      patientWrite,
      encounterRead,
      encounterWrite,
      observationRead,
      observationWrite,
      conditionRead,
      conditionWrite,
      medicationRead,
      medicationWrite,
      allergyRead,
      allergyWrite,
      conflictRead,
      conflictResolve,
      alertRead,
      alertWrite,
      syncRead,
      formularyRead,
      anchorRead,
      supplyRead,
      smartLaunch,
      consentRead,
      consentWrite,
    ],
    roleSiteAdmin: [
      patientRead,
      patientWrite,
      encounterRead,
      encounterWrite,
      observationRead,
      observationWrite,
      conditionRead,
      conditionWrite,
      medicationRead,
      medicationWrite,
      allergyRead,
      allergyWrite,
      conflictRead,
      conflictResolve,
      alertRead,
      alertWrite,
      syncRead,
      syncTrigger,
      formularyRead,
      formularyWrite,
      anchorRead,
      anchorTrigger,
      supplyRead,
      supplyWrite,
      deviceManage,
      smartLaunch,
      smartRegister,
      consentRead,
      consentWrite,
    ],
    roleRegionalAdmin: [
      patientRead,
      patientWrite,
      encounterRead,
      encounterWrite,
      observationRead,
      observationWrite,
      conditionRead,
      conditionWrite,
      medicationRead,
      medicationWrite,
      allergyRead,
      allergyWrite,
      conflictRead,
      conflictResolve,
      alertRead,
      alertWrite,
      syncRead,
      syncTrigger,
      formularyRead,
      formularyWrite,
      anchorRead,
      anchorTrigger,
      supplyRead,
      supplyWrite,
      deviceManage,
      smartLaunch,
      smartRegister,
      consentRead,
      consentWrite,
    ],
  };

  /// Returns `true` if [role] includes the given [permission].
  static bool hasPermission(String role, String permission) {
    final perms = rolePermissions[role];
    if (perms == null) return false;
    return perms.contains(permission);
  }

  /// Returns the display-friendly name for a role.
  static String roleDisplayName(String role) {
    switch (role) {
      case roleCHW:
        return 'Community Health Worker';
      case roleNurse:
        return 'Nurse';
      case rolePhysician:
        return 'Physician';
      case roleSiteAdmin:
        return 'Site Administrator';
      case roleRegionalAdmin:
        return 'Regional Administrator';
      default:
        return role;
    }
  }
}

File: open-nucleus-app/src/lib/permissions.ts
/**
 * RBAC permission constants mirroring the Go backend (internal/model/rbac.go).
 */
export const Permission = {
  PatientRead: 'patient:read',
  PatientWrite: 'patient:write',
  EncounterRead: 'encounter:read',
  EncounterWrite: 'encounter:write',
  ObservationRead: 'observation:read',
  ObservationWrite: 'observation:write',
  ConditionRead: 'condition:read',
  ConditionWrite: 'condition:write',
  MedicationRead: 'medication:read',
  MedicationWrite: 'medication:write',
  AllergyRead: 'allergy:read',
  AllergyWrite: 'allergy:write',
  ConflictRead: 'conflict:read',
  ConflictResolve: 'conflict:resolve',
  AlertRead: 'alert:read',
  AlertWrite: 'alert:write',
  SyncRead: 'sync:read',
  SyncTrigger: 'sync:trigger',
  FormularyRead: 'formulary:read',
  FormularyWrite: 'formulary:write',
  AnchorRead: 'anchor:read',
  AnchorTrigger: 'anchor:trigger',
  SupplyRead: 'supply:read',
  SupplyWrite: 'supply:write',
  SmartLaunch: 'smart:launch',
  SmartRegister: 'smart:register',
  DeviceManage: 'device:manage',
  ConsentRead: 'consent:read',
  ConsentWrite: 'consent:write',
} as const;

export type PermissionValue = (typeof Permission)[keyof typeof Permission];

/**
 * Role constants mirroring the Go backend.
 */
export const Role = {
  CommunityHealthWorker: 'community_health_worker',
  Nurse: 'nurse',
  Physician: 'physician',
  SiteAdministrator: 'site_administrator',
  RegionalAdministrator: 'regional_administrator',
} as const;

export type RoleValue = (typeof Role)[keyof typeof Role];

/**
 * Maps each role to its allowed permissions (mirrors Go RolePermissions).
 */
export const RolePermissions: Record<RoleValue, PermissionValue[]> = {
  [Role.CommunityHealthWorker]: [
    Permission.PatientRead,
    Permission.ObservationRead,
    Permission.ObservationWrite,
    Permission.AlertRead,
    Permission.SyncRead,
  ],
  [Role.Nurse]: [
    Permission.PatientRead,
    Permission.EncounterRead,
    Permission.EncounterWrite,
    Permission.ObservationRead,
    Permission.ObservationWrite,
    Permission.MedicationRead,
    Permission.ConditionRead,
    Permission.AllergyRead,
    Permission.AlertRead,
    Permission.SyncRead,
  ],
  [Role.Physician]: [
    Permission.PatientRead,
    Permission.PatientWrite,
    Permission.EncounterRead,
    Permission.EncounterWrite,
    Permission.ObservationRead,
    Permission.ObservationWrite,
    Permission.ConditionRead,
    Permission.ConditionWrite,
    Permission.MedicationRead,
    Permission.MedicationWrite,
    Permission.AllergyRead,
    Permission.AllergyWrite,
    Permission.ConflictRead,
    Permission.ConflictResolve,
    Permission.AlertRead,
    Permission.AlertWrite,
    Permission.SyncRead,
    Permission.FormularyRead,
    Permission.AnchorRead,
    Permission.SupplyRead,
    Permission.SmartLaunch,
    Permission.ConsentRead,
    Permission.ConsentWrite,
  ],
  [Role.SiteAdministrator]: [
    Permission.PatientRead,
    Permission.PatientWrite,
    Permission.EncounterRead,
    Permission.EncounterWrite,
    Permission.ObservationRead,
    Permission.ObservationWrite,
    Permission.ConditionRead,
    Permission.ConditionWrite,
    Permission.MedicationRead,
    Permission.MedicationWrite,
    Permission.AllergyRead,
    Permission.AllergyWrite,
    Permission.ConflictRead,
    Permission.ConflictResolve,
    Permission.AlertRead,
    Permission.AlertWrite,
    Permission.SyncRead,
    Permission.SyncTrigger,
    Permission.FormularyRead,
    Permission.FormularyWrite,
    Permission.AnchorRead,
    Permission.AnchorTrigger,
    Permission.SupplyRead,
    Permission.SupplyWrite,
    Permission.DeviceManage,
    Permission.SmartLaunch,
    Permission.SmartRegister,
    Permission.ConsentRead,
    Permission.ConsentWrite,
  ],
  [Role.RegionalAdministrator]: [
    Permission.PatientRead,
    Permission.PatientWrite,
    Permission.EncounterRead,
    Permission.EncounterWrite,
    Permission.ObservationRead,
    Permission.ObservationWrite,
    Permission.ConditionRead,
    Permission.ConditionWrite,
    Permission.MedicationRead,
    Permission.MedicationWrite,
    Permission.AllergyRead,
    Permission.AllergyWrite,
    Permission.ConflictRead,
    Permission.ConflictResolve,
    Permission.AlertRead,
    Permission.AlertWrite,
    Permission.SyncRead,
    Permission.SyncTrigger,
    Permission.FormularyRead,
    Permission.FormularyWrite,
    Permission.AnchorRead,
    Permission.AnchorTrigger,
    Permission.SupplyRead,
    Permission.SupplyWrite,
    Permission.DeviceManage,
    Permission.SmartLaunch,
    Permission.SmartRegister,
    Permission.ConsentRead,
    Permission.ConsentWrite,
  ],
};

/** Check if a role includes a specific permission. */
export function hasPermission(role: string, permission: string): boolean {
  const perms = RolePermissions[role as RoleValue];
  if (!perms) return false;
  return perms.includes(permission as PermissionValue);
}
